Various improvements in std::collections docs

lukaramu opened the pull request:
The meat of this PR are:
* changes to (almost all?) iterator struct docs in std::collections such that they use the standard iterator boilerplate and state where they are created
* a bunch of added links (at least as much as possible given std::collections mostly being a facade and whatnot :sweat_smile:)
* an example for `Bound`
* changed phrasing for some summary sentences to be less redundant as well as more consistant with others in the module

There also are various other fixes, e.g. removing parens from method names in the module docs, changing some imperatives to 3rd person, etc.

r? @steveklabnik

frewsxcv commented on src/libcollections/binary_heap.rs:
```diff
@@ -218,10 +219,14 @@ pub struct BinaryHeap<T> {
 
-/// A container object that represents the result of the [`peek_mut`] method
-/// on `BinaryHeap`. See its documentation for details.
+/// Object representing a mutable reference to the greatest item on a
```
I think we try to avoid using the term 'object' to describe structures since Rust isn't really OO. Personally I use 'structure', though there might be a better word.

frewsxcv replied:
Maybe "Structure wrapping a mutable reference..."?

lukaramu replied:
Yeah, that sounds better!

frewsxcv commented on src/libcollections/lib.rs:
```diff
@@ -135,6 +135,23 @@ mod std {
+/// map.insert(5, "b");
+/// map.insert(8, "c");
+///
+/// for (key, value) in map.range((Excluded(3), Included(8))) {
```
I think you should be able to use the `Range` sugar here: `4..9`

lukaramu replied:
Yeah, that's true. I adapted this example from the one for [`BTreeMap::range`], which I agree is not ideal; as far as I could tell those kinds of methods are the only place in the (stable) standard library where `Bound` can be used right now though. I'd rather use something like the examples in [`RangeArgument`], but that's not stable yet so I wasn't sure if that is ok to use.

[`BTreeMap::range`]: https://doc.rust-lang.org/nightly/std/collections/struct.BTreeMap.html#method.range
[`RangeArgument`]: https://doc.rust-lang.org/nightly/std/collections/range/trait.RangeArgument.html#tymethod.start

lukaramu commented:
Linkchecker is failing on the links I added to `HashSet`. I thought it should work since it does for `HashMap`, but upon investigation it seems like [it's whitelisted here][0] (see #32130).

Should I add it to the whitelist, or should I remove the links there again (and in other places too)? I feel like the docs are better with the links there in the version directly in `std::collections`, but at the same time, it's unfortunate that the version that's within `std::collections::hash_set` has broken links...

[0]: https://github.com/rust-lang/rust/blob/96d35947c36bba6c6545c06768b5358c0f8bf3dc/src/tools/linkchecker/main.rs#L141

frewsxcv commented:
> Should I add it to the whitelist, or should I remove the links there again (and in other places too)? I feel like the docs are better with the links there in the version directly in std::collections, but at the same time, it's unfortunate that the version that's within std::collections::hash_set has broken links...

Yeah, it's a bit of an unfortunate situation right now with docs links, though hopefully it will be fixed in the future. Either whitelisting or removing the links is fine with me. Do whichever your prefer!

steveklabnik commented on src/libcollections/lib.rs:
```diff
@@ -135,6 +135,23 @@ mod std {
 }
 
 /// An endpoint of a range of keys.
+/// # Examples
```
there should be an extra `///` above this line

lukaramu commented:
I went with adding `collections/hash_set/struct.HashSet.html` to the whitelist.

After talking about it on IRC, I also added a second example for `Bound`. It uses the unstable `RangeArgument` trait, but I think it's the more illustrative one; I placed it first accordingly. I also added a note the existing example, that should address the comment about being able to use range sugar.

frewsxcv commented:
Wooo thanks! ✨ 

@bors r+ rollup

bors commented:
:pushpin: Commit 2a23e6e has been approved by `frewsxcv`

<!-- @bors r=frewsxcv 2a23e6e277bc6a6a00c5e5d9d5a159265bef8c24 -->

bors commented:
:hourglass: Testing commit 2a23e6e with merge 5f13a3b...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2933), [status-travis](https://travis-ci.org/rust-lang/rust/builds/222283086)
Approved by: frewsxcv
Pushing 5f13a3b540ab6024665322d716e487c800645f24 to master...
